File: Antlr/src/main/scala/io/github/chaosunity/casc/parsing/statement/RangedForStatement.scala
package io.github.chaosunity.casc.parsing.statement

import io.github.chaosunity.casc.exceptions.UnsupportedRangedTypeException
import io.github.chaosunity.casc.parsing.`type`.{BuiltInType, Type}
import io.github.chaosunity.casc.parsing.expression.Expression
import io.github.chaosunity.casc.parsing.scope.Scope
import io.github.chaosunity.casc.util.TypeChecker


class RangedForStatement(val iteratorVariableStatement: Statement,
                         val startExpression: Expression,
                         val forType: ForType.Value,
                         val endExpression: Expression,
                         val statement: Statement,
                         val iteratorVariableName: String,
                         val scope: Scope) extends ForStatement {
    if (!TypeChecker.isInteger(startExpression.`type`) &&
        !TypeChecker.isInteger(endExpression.`type`))
        throw new UnsupportedRangedTypeException(startExpression.`type`, endExpression.`type`)

    def `type`: Type = startExpression.`type`
}

File: src/main/kotlin/io/github/chaosunity/casc/parsing/node/statement/RangedForStatement.kt
package io.github.chaosunity.casc.parsing.node.statement

import io.github.chaosunity.casc.parsing.node.expression.Expression
import io.github.chaosunity.casc.parsing.scope.Scope

data class RangedForStatement(
    val iteratorVariable: Statement<*>,
    val startExpression: Expression<*>,
    val arrowText: String,
    val endExpression: Expression<*>,
    val statement: Statement<*>,
    val iteratorVarName: String,
    val scope: Scope
) : ForStatement<RangedForStatement>() {
    init {
        if (!startExpression.isInt() || !endExpression.isInt()) throw RuntimeException("For statement only supports integer range.s")
    }

    val stopTo = if (arrowText.endsWith('>')) arrowText.startsWith('-') else arrowText.endsWith('-')
    val forward = arrowText.endsWith('>')
}